File: daily/Binary Tree/124-Binary-tree-maximum-path-sum.go
package daily

import "math"

func maxPathSum(root *TreeNode) int {
	bestSum := math.MinInt
	var dfs func(node *TreeNode) int
	dfs = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		left := dfs(node.Left)
		right := dfs(node.Right)
		if left < 0 {
			left = 0
		}
		if right < 0 {
			right = 0
		}

		bestSum = max(bestSum, node.Val+left+right)
		return node.Val + max(left, right)
	}
	dfs(root)
	return bestSum
}

File: daily/Binary Tree/124-Binary-tree-maximum-path-sum.py
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.best = float('-inf')
        def dfs(node):
            if not node:
                return 0
            left = dfs(node.left)
            right = dfs(node.right)
            if left < 0:
                left = 0
            if right < 0:
                right = 0
            self.best = max(self.best, node.val + left + right)
            return node.val + max(left, right)
        dfs(root)
        return self.best